src/core/thread/discover_scanner.cpp

```diff
@@ -247,14 +248,21 @@ void DiscoverScanner::HandleDiscoverComplete(void)
             mShouldRestorePanId = false;
         }
 
-        mState = kStateIdle;
+        // Post the tasklet to change `mState` and invoke handler
+        // callback. This allows users to safely call OT APIs from
+        // the callback.
+        mScanDoneTask.Post();
+        break;
+    }
+}
 
-        if (mHandler)
-        {
-            mHandler(nullptr, mHandlerContext);
```

Comment:
Tasklet is not necessary as long as we can set `mHandler = nullptr` before calling?

Reply:
Not correct. The issue is not that `mHandler` is not null.
Generally the callbacks are invoked as the last step, but in the current design of "discover scan" this is not the case and there may be addition processing happening after the callback is invoked. Please check the https://github.com/openthread/openthread/issues/8110 for more details.

Reply:
Thanks for explanation.

Reply:
Sure. I think we can later change the discover scan design to avoid this situation (ensure callback is called at the end) but that's beyond the  scope of this PR. We can take care of it later.